plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs

```diff
@@ -0,0 +1,32 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module UntypedPlutusCore.Check.Scope (checkScope) where
+
+import Control.Lens hiding (index)
+import UntypedPlutusCore.Core.Type as UPLC
+import UntypedPlutusCore.DeBruijn as UPLC
+
+checkScope :: forall name uni fun a. HasIndex name
+           => UPLC.Term name uni fun a
+           -> Bool
+checkScope = \case
```

Comment:
I looked at this and thought "can't you just replace this with `withL 0`? But then we'd allow a top level `Var 0`, which we don't want. 

Alternative: just use one function and add a clause to the `Var` case that says `&& index > 0`. I think I'd dislike that less.

Reply:
the `&& index > 0` would makeit potentially slower (slightly?), and this pass we will run after every script deserialization.

Reply:
True, but in a pretty trivial way. We can do it this way and then benchmark it :) Doing performance optimizations without numbers is risky.